File: pie_sol/pie_sol/utils/js/customer.js
frappe.ui.form.on("Customer", {
    refresh:function(frm){
                if (cur_frm.fields_dict.custom_records) {
            // Hide the "Add Multiple Rows" button
                    cur_frm.fields_dict['custom_records'].$wrapper.find(".grid-add-row").hide();
                };
            

            frm.fields_dict['custom_records'].grid.get_field('brand_name').get_query = function(doc, cdt, cdn) {
                return {
                    filters: [
                        ['Item Group', 'is_group', '=', 1],
                        ['Item Group', 'parent_item_group', '=', 'All Item Groups'],
                     
                    ]
                };
            };
            frm.fields_dict['custom_records'].grid.get_field('mc_type').get_query = function(doc, cdt, cdn) {
                let row = locals[cdt][cdn];
                return {
                    filters: [
                        ['Item Group', 'is_group', '=', 1],
                        ['Item Group', 'parent_item_group', '=', row.brand_name],
                     
                    ]
                };
            };
            frm.fields_dict['custom_records'].grid.get_field('model_name').get_query = function(doc, cdt, cdn) {
                let row = locals[cdt][cdn];
                return {
                    filters: [
                        ['Item Group', 'is_group', '=', 0],
                        ['Item Group', 'parent_item_group', '=', row.mc_type],
                     
                    ]
                };
            };
            frm.fields_dict['custom_records'].grid.get_field('indoor_model').get_query = function(doc, cdt, cdn) {
                let row = locals[cdt][cdn];
                return {
                    filters: [
                        ['Item', 'item_group', '=', row.brand_name],
                        ['Item', 'custom_mc_type', '=', row.mc_type],
                        ['Item', 'custom_model', '=', row.model_name],
                    ]
                };
            };
            frm.fields_dict['custom_records'].grid.get_field('outdoor_model').get_query = function(doc, cdt, cdn) {
                let row = locals[cdt][cdn];
                return {
                    filters: [
                        ['Item', 'item_group', '=', row.brand_name],
                        ['Item', 'custom_mc_type', '=', row.mc_type],
                        ['Item', 'custom_model', '=', row.model_name],
                    ]
                };
            };
        } ,








    // customer_name: function(frm) {
    //     var abbrev_name = function(str1) {
    //         var split_names = str1.trim().split(" ");

    //         if (split_names.length > 1) {
    //             return (split_names[0].charAt(0) + (split_names.length > 2 ? split_names[1].charAt(0) + split_names[2].charAt(0) : split_names[1].charAt(0) + split_names[1].charAt(1)));
    //         }
    //         return split_names[0].charAt(0) + split_names[0].charAt(1) + split_names[0].charAt(2);
    //     };

    //     var customer_name = frm.doc.customer_name;

    //     var abbreviation = abbrev_name(customer_name);

    //     var abbreviation = abbreviation.toUpperCase();
    //     frm.set_value('custom_abbreviation', abbreviation);
    // },


    // custom_abbreviation: function(frm) {
    //     var abbreviation = frm.doc.custom_abbreviation;
    //     if (abbreviation) {                                                  
    //         frappe.call({
    //             method: 'frappe.client.get_list',
    //             args: {
    //                 doctype: 'Customer',
    //                 filters: {
    //                     'custom_abbreviation': frm.doc.custom_abbreviation
    //                 },
    //                 fields: ['name']
    //             },
    //             callback: function(response) {
    //                 console.log(response.message)
    //                 if (response.message && response.message.length >=1) {
    //                     frm.fields_dict.custom_abbreviation.set_description('<span style="color:red;">' +__('User abbreviation already exists kindly change customer name')+'</span>');
    //                 } else {
    //                     frm.fields_dict.custom_abbreviation.set_description('');
    //                 }
    //             }
    //         });
    //     }
    // },



    
    custom_add_record:function(frm){
        let d = new frappe.ui.Dialog({
            title: 'Enter Machine details',
            fields: [
                {
                    label: 'Brand Name',
                    fieldname: 'add_brand_name',
                    fieldtype: 'Link',
                    options: 'Item Group', 
                    reqd: true,
                    get_query: function() {
                        return {
                        filters: {


      				            is_group:1,
                                parent_item_group:'All Item Groups',
                            }
                        };
                    }
                },
                {
                    label: 'MC Type',
                    fieldname: 'add_mc_type',
                    fieldtype: 'Link',
                    options: 'Item Group',
                    reqd: true, 
                    get_query: function() {
                        return {

                        filters: {


                                parent_item_group:d.get_value('add_brand_name'),
      				            is_group:1,
                            }
                        };
                    }
                },
                {
                    label: 'Model Name',
                    fieldname: 'add_model_name',
                    fieldtype: 'Link',
                    options: 'Item Group',
                    reqd: true, 
                    get_query: function() {
                        
                        return {
                            
                        filters: {


      				            is_group:0,
                                parent_item_group:d.get_value('add_mc_type'),
                            }
                        };
                    }
                },
                {
                    label: 'Tonnes',
                    fieldname: 'add_tonnes',
                    fieldtype: 'Select',
                    options:['1','1.5','2','2.5','3','3.5','4','4.5','5','5.5'],
                    reqd: true
                },
                {
                    label: 'Indoor model',
                    fieldname: 'add_indoor_model',
                    fieldtype: 'Link',
                    options: 'Item',
                    reqd: true, 
                    get_query: function() {
                        
                        return {
                            
                        filters: {
                                item_group:d.get_value('add_brand_name'),
                                custom_mc_type:d.get_value('add_mc_type'),
                                custom_model:d.get_value('add_model_name')
                            }
                        };
                    }
                },
                {
                    label: 'Outdoor model',
                    fieldname: 'add_outdoor_model',
                    fieldtype: 'Link',
                    options: 'Item',
                    reqd: true, 
                    get_query: function() {
                        
                        return {
                            
                        filters: {
                            item_group:d.get_value('add_brand_name'),
                            custom_mc_type:d.get_value('add_mc_type'),
                            custom_model:d.get_value('add_model_name')
                            }
                        };
                    }
                },
                {
                    label: 'Date',
                    fieldname: 'add_date',
                    fieldtype: 'Date',
                    reqd: true
                }
            ],
            size: 'small', // small, large, extra-large 
            primary_action_label: 'Submit',
            primary_action(values) {


                

            // Create a new Machine document
                    frappe.call({
                        method: 'pie_sol.pie_sol.utils.py.machine.create_machine',
                        args: {
                    
                        customer_name:frm.doc.customer_name,
                        brand_name: values.add_brand_name,
                        mc_type: values.add_mc_type,
                        model_name: values.add_model_name,
                        indoor_model: values.add_indoor_model,
                        outdoor_model: values.add_outdoor_model,
                        tonnes: values.add_tonnes,
                        date:values.add_date
                        },
                    callback: function(response) {
                        frm.refresh_fields('custom_records')
                    
                   
                }
            }); 
       
                d.hide();
            }
        });
        
        d.show();
        
    },
                

});











frappe.ui.form.on("Service Records", {


    refresh(frm) {
      frm.set_query("brand_name", function () {
          return {
              filters: {
                  is_group: 1,
              },
          };
      });
    },

});

File: pie_sol/pie_sol/utils/py/customer.py
import frappe
import time
from frappe import _

def validate_customer_abbreviation(doc, events):
    if doc.is_new():
        if doc.custom_abbreviation:
            count=frappe.db.get_list('Customer',filters={
                'custom_abbreviation':doc.custom_abbreviation
            })
            if len(count)!=0:
                existing_customer = frappe.db.exists('Customer', {'custom_abbreviation': doc.custom_abbreviation})

                frappe.throw(_("Abbreviation '{}' already exists for customer '{}'").format(doc.custom_abbreviation, existing_customer), title=_('Duplicate Abbreviation'))



def mobile_number_validation1(doc,events):
    num=doc.custom_mobile_number1
    if num.isdigit():
        if len(num)!=10:
            frappe.throw(_("Mobile must have 10 digit"))
    else:
        frappe.throw(_("Invalid mobile format"))

        
def mobile_number_validation2(doc,events):
    if doc.custom_mobilenumber2:
        num=doc.custom_mobilenumber2
        if num.isdigit():
            if len(num)!=10:
                frappe.throw(_("Mobile must have 10 digit"))
        else:
            frappe.throw(_("Invalid mobile format"))



def validate_customer_emails(doc, method):
    for row in doc.get("custom_information"):
        if row.email:
            frappe.utils.validate_email_address(row.email, True)
#############

def customerservice(doc,events):
    customer_name=doc.customer_name
    cus=frappe.get_doc('Customer',customer_name)
    cus.append("custom_records",
        {
            'machine':doc.name,
            'brand_name':doc.brand_name,
            'mc_type':doc.mc_type,
            'model_name':doc.model_name,
            'indoor_model':doc.indoor_model,
            'outdoor_model':doc.outdoor_model,
            'tonnes':doc.tonnes,
            'date':doc.installation_date,

        })
    cus.save()
    cus.reload()


def duplicateEntry(doc , events):
    if doc.custom_records:
        entries = []
        for i in doc.custom_records:
            if i.machine in entries:
                frappe.throw(i.machine + " Already in available in table")
            else:
                entries.append(i.machine)


# def updaterecords(doc,events):
#     name=doc.name
#     customer_name=doc.customer_name
#     cus=frappe.get_doc('Customer',customer_name)
#     for row in cus.custom_records:
#         # if row.machine==name and row.brand_name==doc.brand_name and row.mc_type==doc.mc_type and row.model_name==doc.model_name and row.indoor_model==doc.indoor_model and row.outdoor_model==doc.outdoor_model and row.tonnes==doc.tonnes and row.date==doc.installation_date:
#         #     return
#         # else:

#             row.machine=doc.name
#             row.brand_name=doc.brand_name
#             row.mc_type=doc.mc_type
#             row.model_name=doc.model_name
#             row.indoor_model=doc.indoor_model
#             row.outdoor_model=doc.outdoor_model
#             row.tonnes=doc.tonnes
#             row.date=doc.installation_date
        
#             print("$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$")
#     cus.save()




# def machine_update(doc, events):
#     for row in doc.custom_records:
#         machine = frappe.get_doc("Machine", row.machine)
#         print(machine, machine.brand_name, machine.installation_date, machine.indoor_model, machine.outdoor_model, machine.model_name, machine.mc_type, machine.tonnes)
#         print(machine, row.brand_name, row.date, row.indoor_model, row.outdoor_model, row.model_name, row.mc_type, row.tonnes)
        
#         if frappe.db.get_value("Machine", machine.name, "brand_name") != row.brand_name:
#             machine.brand_name = row.brand_name
#         if frappe.db.get_value("Machine", machine.name, "mc_type") != row.mc_type:
#             machine.mc_type = row.mc_type
#         if frappe.db.get_value("Machine", machine.name, "model_name") != row.model_name:
#             machine.model_name = row.model_name
#         if frappe.db.get_value("Machine", machine.name, "indoor_model") != row.indoor_model:
#             machine.indoor_model = row.indoor_model
#         if frappe.db.get_value("Machine", machine.name, "outdoor_model") != row.outdoor_model:
#             machine.outdoor_model = row.outdoor_model
#         if frappe.db.get_value("Machine", machine.name, "tonnes") != row.tonnes:
#             machine.tonnes = row.tonnes
#         if frappe.db.get_value("Machine", machine.name, "installation_date") != row.date:
#             machine.installation_date = row.date
            
#         machine.save()




def customer_update(doc, events):
    customer = frappe.get_doc("Customer", doc.customer_name)
    for row in customer.custom_records:
        
        if  row.brand_name != doc.brand_name and row.machine==doc.name:
            row.brand_name = doc.brand_name
        if  row.mc_type != doc.mc_type and row.machine==doc.name:
            row.mc_type = doc.mc_type
        if row.model_name != doc.model_name and row.machine==doc.name:
            row.model_name = doc.model_name
        if row.indoor_model!= doc.indoor_model and row.machine==doc.name:
            row.indoor_model = doc.indoor_model
        if row.outdoor_model != doc.outdoor_model and row.machine==doc.name:
            row.outdoor_model=doc.outdoor_model
        if row.tonnes != doc.tonnes and row.machine==doc.name:
            row.tonnes = doc.tonnes
        if row.date != doc.installation_date and row.machine==doc.name:
            row.date = doc.installation_date
            
    customer.save()
    













# def machine_updaterecords(doc, events):
#     for row in doc.custom_records:
#         mac = frappe.get_doc('Machine', row.machine)
#         if mac:
#             # if mac.brand_name==row.brand_name and mac.mc_type==row.mc_type and mac.model_name==row.model_name and mac.indoor_model==row.indoor_model and mac.outdoor_model==row.outdoor_model and mac.tonnes==row.tonnes and mac.installation_date==row.date:
#             #     return
#             # else:
#                 mac.update({
#                     'brand_name': row.brand_name,
#                     'mc_type': row.mc_type,
#                     'model_name': row.model_name,
#                     'indoor_model': row.indoor_model,
#                     'outdoor_model': row.outdoor_model,
#                     'tonnes': row.tonnes,
#                     'installation_date': row.date
#             })


  
        
